include/cantera/thermo/BinarySolutionTabulatedThermo.h

```diff
@@ -143,16 +191,58 @@ class BinarySolutionTabulatedThermo : public IdealSolidSolnPhase
         return "BinarySolutionTabulatedThermo";
     }
 
+    virtual bool addSpecies(shared_ptr<Species> spec);
     virtual void initThermo();
     virtual void getParameters(AnyMap& phaseNode) const;
     virtual void initThermoXML(XML_Node& phaseNode, const std::string& id_);
 
+    /**
+     * returns an array of partial molar volumes of the species
+     * in the solution. Units: m^3 kmol-1.
+     *
+     * The partial molar volumes are derived as shown in the equations in the detailed
+     * description section.
+     *
+     * @param vbar  Output vector of partial molar volumes. Length: m_kk.
+     */
+    virtual void getPartialMolarVolumes(doublereal* vbar) const;
+
+    /**
+     * Overloads the calcDensity() method of IdealSolidSoln to also consider non-ideal
+     * behavior.
+     *
+     * The formula for this is
+     *
+     * \f[
+     * \rho = \frac{\sum_k{X_k W_k}}{V_\mathrm{m}}
+     * \f]
+     *
+     * where \f$X_k\f$ are the mole fractions, \f$W_k\f$ are the molecular weights, and
+     * \f$V_\mathrm{m}\f$ is the molar volume interpolated from \f$V_{\mathrm{m,tab}}\f$.
+     */
+    virtual void calcDensity();
+
 protected:
     //! If the compositions have changed, update the tabulated thermo lookup
     virtual void compositionChanged();
 
-    //! Species thermodynamics interpolation functions
-    std::pair<double,double> interpolate(double x) const;
+    //! Species thermodynamics linear interpolation function
+    /*!
+     *  Tabulated values are only interpolated within the limits of the provided mole
+     *  fraction. If these limits are exceeded, the values are capped at the lower or
+     *  the upper limit.
+     */
+    double interpolate(double x, const vector_fp& molefrac,
+                       const vector_fp& inputData) const;
+
+    //! Numerical derivation of the molar volume table
+    /*!
+     *  Tabulated values are only interpolated within the limits of the provided mole
+     *  fraction. If these limits are exceeded, the values are capped at the lower or
+     *  the upper limit.
+     */
+    void derive(vector_fp& inputData, vector_fp& derivedData,
+                vector_fp& moleFraction);
```

Comment:
A few comments on this function:
- This is the numerical _derivative_ of the molar volume
- I think `diff` would be a better name for this function is used, the mole fraction vector is `m_molefrac_tab`
- I don't think the `moleFraction` argument is needed -- the only way this function is called, 
- The usual order of arguments would be to have the input arguments (`inputData` and `moleFraction` first, followed by the output argument.
- The input arguments should be marked as `const`
- The function itself is can be `const` as well.